### PR TITLE
Core Data: Resolve user capabilities when fetching an entity

### PR DESCRIPTION
**Pull request by Mamaduka**
## What?
Inspired by https://github.com/WordPress/gutenberg/pull/43480#issuecomment-1231810231.
Hopefully, this will be more effective after merging #63415 and #63423.

PR updates the `getEntityRecord` to resolve entity-related user permission when fetching entity data.

## Why?
Most of the time, actions to interact with the entity are displayed after loading it. Resolving user capability selectors for the same entity can improve the UI (no more popcorning) and reduce the number of requests the app makes.

### Manually run selectors
1. Open a post or page.
2. Run selectors below in order using DevTools Console.
3. Running the `canUser` selector for the same entity shouldn't trigger a REST API request.

```
// 1. Fetch media details.
wp.data.select( 'core' ).getEntityRecord( 'root', 'media', 1342 );

// 2. Fetch user permission related to the media. It shouldn't trigger a REST API request.
wp.data.select( 'core' ).canUser( 'update', { kind: 'root', name: 'media', id: 1342 } );
```

### Pattern blocks

1. Open a post or page.
2. Add multiple synced pattern blocks.
3. Save and reload the editor.
4. Confirm using Network tools that no `OPTIONS` requests are made for these patterns. You can use `v2/blocks` to filter the endpoint.

### Testing Instructions for Keyboard
Same.

## Screenshots or screencast 
**Requests before with four pattern blocks on canvas:**
CleanShot 2024-07-15 at 16 52 13

**Requests after:**
CleanShot 2024-07-15 at 16 52 36

### CODE REVIEW

File: packages/core-data/src/resolvers.js [hunk 1]
```diff
@@ -165,8 +165,44 @@ export const getEntityRecord =
...
+						'create',
+						'read',
+						'update',
+						'delete',
+					] ) {
+						const permissionKey = [ action, kind, name, key ]
+							.filter( Boolean )
+							.join( '/' );
+
+						dispatch.receiveUserPermission(
```
Comment (youknowriad) [inline packages/core-data/src/resolvers.js:187]: I think that's a good idea but we should also mark the resolvers as resolved to avoid further REST API requests when the selectors get called no?

Comment (Mamaduka) [reply packages/core-data/src/resolvers.js:187]: Correct, and I think I need to update `canUser` as well.

I see that some resolvers for similar cases call both `startResolution` and `finishResolution` actions, while some only call `finishResolution`. Which method is generally preferred?

Comment (jsnajdr) [reply packages/core-data/src/resolvers.js:187]: If you'll do it inside the `batch`, then only `finishResolution` makes sense. `start` + `finish` is useful is when you want the observers to see the "tick", but when calling listeners is disables, as it is in `batch`, then `start` does nothing useful.

Comment (Mamaduka) [reply packages/core-data/src/resolvers.js:187]: Thanks, @jsnajdr!

File: packages/core-data/src/resolvers.js [hunk 2]
```diff
@@ -165,8 +165,44 @@ export const getEntityRecord =
...
+					read: 'GET',
+					update: 'PUT',
+					delete: 'DELETE',
+				};
+				for ( const [ actionName, methodName ] of Object.entries(
+					methods
+				) ) {
+					permissions[ actionName ] =
+						allowedMethods.includes( methodName );
+				}
```
Comment (tyxla) [inline packages/core-data/src/resolvers.js]: This is getting a bit repetitive. I'm sure we're using the same logic in `canUser()`. Can we deduplicate?

File: packages/core-data/src/resolvers.js [hunk 3]
```diff
@@ -165,8 +165,44 @@ export const getEntityRecord =
...
+
+				registry.batch( () => {
+					dispatch.receiveEntityRecords( kind, name, record, query );
+
+					for ( const action of [
+						'create',
+						'read',
+						'update',
+						'delete',
```
Comment (tyxla) [inline packages/core-data/src/resolvers.js]: Those will be redeclared more than once, too. Should we deduplicate?

File: packages/core-data/src/resolvers.js (hunk 2 above)
Comment (Mamaduka) [reply packages/core-data/src/resolvers.js]: That's the plan. I will probably extract those into new utility methods.

Work in progress names:
* `getCapabilitiesFromReponseHeaders` - handles this particular case.
* `getNormalizedCapabilityKey` - handles key generation.

File: packages/core-data/src/resolvers.js (hunk 3 above)
Comment (Mamaduka) [reply packages/core-data/src/resolvers.js]: +1. I will extract this into a constant.

File: packages/core-data/src/resolvers.js (hunk 2 above)
Comment (Mamaduka) [reply packages/core-data/src/resolvers.js]: @tyxla, what do you think about this deduplication - 27b4b0d18db2797e37a32a6970dcc779b6a046e9?

File: packages/block-library/src/block/edit.js [hunk 4]
```diff
@@ -256,27 +292,18 @@ function ReusableBlockEdit( {
...
+					recordId={ ref }
+					canOverrideBlocks={ canOverrideBlocks }
+					hasContent={ !! content }
+					handleEditOriginal={
+						onNavigateToEntityRecord
+							? handleEditOriginal
+							: undefined
+					}
+					resetContent={ resetContent }
+				/>
```
Comment (Mamaduka) [inline packages/block-library/src/block/edit.js:306]: Unrelated change showcases the benefits of this proposal. I am happy to extract it into a separate PR.

**Requests before with four pattern blocks on canvas:**
CleanShot 2024-07-15 at 16 52 13


**Requests after:**
CleanShot 2024-07-15 at 16 52 36

**Comment by Mamaduka** [comment]
I think PR is in good shape for reviews and feedback.

File: packages/core-data/src/resolvers.js [hunk 5]
```diff
@@ -165,8 +171,29 @@ export const getEntityRecord =
 					}
 				}
 
-				const record = await apiFetch( { path } );
-				dispatch.receiveEntityRecords( kind, name, record, query );
+				const response = await apiFetch( { path, parse: false } );
```
Comment (jsnajdr) [inline packages/core-data/src/resolvers.js:174]: Setting `parse: false` will disable the `fetchAllMiddleware`. See how it [returns early](https://github.com/WordPress/gutenberg/blob/c4332b60ee79235e76bf6926e8368a72b62b6284/packages/api-fetch/src/middlewares/fetch-all-middleware.js#L78-L81) when `options.parse` is `false`. Is this a concern here?

`fetchAllMiddleware` is relevant for fetching collections, and `getEntityRecord` fetches only one item, so it shouldn't be relevant. But what if I do:
```js
getEntityRecord( 'postType', 'post', '', { per_page: -1 } );
```
This will fetch the entire collection anyway, won't it? And should trigger `fetchAllMiddleware`.

Comment (Mamaduka) [reply packages/core-data/src/resolvers.js:174]: I also had similar concerns during the original discussion: https://github.com/WordPress/gutenberg/pull/43480#issuecomment-1235459202. But it shouldn't affect the `getEntityRecord` resolver.

The `getEntityRecord` supports only fetching a single entity. Any other usage is incorrect, and it's up to consumers to fix it.

Calling `wp.data.select( 'core' ).getEntityRecord( 'postType', 'post', '', { per_page: -1 } );` will make a request, but then will return `undefined`.

File: packages/core-data/src/utils/user-permissions.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	'read',
+	'update',
+	'delete',
+];
+
+export function getUserPermissionsFromResponse( response ) {
+	const permissions = {};
+
+	// Optional chaining operator is used here because the API requests don't
+	// return the expected result in the native version. Instead, API requests
```
Comment (tyxla) [inline packages/core-data/src/utils/user-permissions.js]: Made me re-read it to understand we're talking about the React Native implementation. 

```suggestion
	// return the expected result in the React native version. Instead, API requests
```

File: packages/core-data/src/resolvers.js (hunk 2 above)
Comment (tyxla) [reply packages/core-data/src/resolvers.js]: Looks good, thanks! As long as it's not a public API, we can be a bit more flexible and not super critical about it.

Comment (Mamaduka) [reply packages/core-data/src/resolvers.js]: Yes. These are private utilities.

File: packages/core-data/src/utils/user-permissions.js (hunk 6 above)
Comment (jsnajdr) [reply packages/core-data/src/utils/user-permissions.js]: The React Native special case was discussed in detail here: https://github.com/WordPress/gutenberg/pull/39144#issuecomment-1056701720

In short, RN response always have a parsed body and never include headers. Except when the response is an error.

Another thing that's hard to understand is this line:
```js
const allowedMethods = allowHeader?.allow || allowHeader || ''
```
In what case could `allowHeader` be an object with an `allow` field? This was added by @tyxla in a Lodash removal PR: https://github.com/WordPress/gutenberg/pull/43117/files#diff-fec6390e24332dec4a0c7c74f9f7b9a015ce191e0e6348361ead793ceda3991dR319

The original code was `includes( allowHeader, method )`. I still don't understand accessing the `.allow` field.

Comment (tyxla) [reply packages/core-data/src/utils/user-permissions.js]: This appears to be a mistake 🤔 Which seems to be additionally covered by a pre-existing test fixture. Let me fix.

Comment (tyxla) [reply packages/core-data/src/utils/user-permissions.js]: Fix in #63615

Comment (tyxla) [reply packages/core-data/src/utils/user-permissions.js]: Landed. Feel free to rebase @Mamaduka

Comment (Mamaduka) [reply packages/core-data/src/utils/user-permissions.js]: Done ✅

**Comment by Mamaduka** [comment]
@jsnajdr, @tyxla, any remaining blocks here?

**Review by jsnajdr — APPROVED** [review]
Looks good to go :+1:
